### PR TITLE
[#IOCIT-340] Add jwt verify middleware

### PR DESCRIPTION
**Pull request by Garma00**
#### List of Changes

* [Add jwt verify middleware](https://github.com/pagopa/io-functions-lollipop/commit/8b4edefecaecfb00d8e90d0d7302d4924342a2b9);

#### Motivation and Context

A middleware that verify the jwt is needed.

#### How Has This Been Tested?

Unit test 

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: utils/auth_jwt.ts [hunk 1]
```diff
@@ -62,3 +69,36 @@ export const getValidateAuthJWT = ({
...
+ * @param jwtConfig the config for the jwt
+ * */
+
+export const verifyJWTMiddleware = (
+  jwtConfig: JWTConfig
+): IRequestMiddleware<"IResponseErrorValidation", AuthJWT> => (
+  req
+): Promise<E.Either<IResponseErrorValidation, AuthJWT>> =>
+  pipe(
+    req.headers["x-pagopa-lollipop-auth"],
```
Comment (BurnedMarshal) [inline utils/auth_jwt.ts:85]: The header name should came from the configuration.
This change will make the `verifyJWTMiddleware` usable to validate any JWT

File: utils/auth_jwt.ts [hunk 2]
```diff
@@ -7,9 +7,16 @@ import { pipe } from "fp-ts/lib/function";
...
+import { IRequestMiddleware } from "@pagopa/io-functions-commons/dist/src/utils/request_middleware";
+import {
+  IResponseErrorValidation,
+  ResponseErrorValidation
+} from "@pagopa/ts-commons/lib/responses";
+import { readableReportSimplified } from "@pagopa/ts-commons/lib/reporters";
 import { AssertionRef } from "../generated/definitions/internal/AssertionRef";
 import { OperationId } from "../generated/definitions/internal/OperationId";
 
+import { LollipopAuthBearer } from "../generated/definitions/external/LollipopAuthBearer";
```
Comment (BurnedMarshal) [inline utils/auth_jwt.ts:19]: The name of this type should be generic, not only for lollipop.
es `JWTAuthBearer`

File: utils/auth_jwt.ts [hunk 3]
```diff
@@ -62,3 +69,36 @@ export const getValidateAuthJWT = ({
...
+      )
+    ),
+    E.map(authBearer => authBearer.replace("Bearer ", "") as NonEmptyString),
+    TE.fromEither,
+    TE.chain(token =>
+      pipe(
+        token,
+        getValidateAuthJWT(jwtConfig),
+        TE.mapLeft(({ message }) =>
+          ResponseErrorValidation("Invalid authJWT", message)
```
Comment (BurnedMarshal) [inline utils/auth_jwt.ts:100]: This should be `Forbidden` instead `ErrorValidation`

File: utils/auth_jwt.ts [hunk 4]
```diff
@@ -62,3 +69,36 @@ export const getValidateAuthJWT = ({
...
+export const verifyJWTMiddleware = (
+  jwtConfig: JWTConfig
+): IRequestMiddleware<"IResponseErrorValidation", AuthJWT> => (
+  req
+): Promise<E.Either<IResponseErrorValidation, AuthJWT>> =>
+  pipe(
+    req.headers["x-pagopa-lollipop-auth"],
+    LollipopAuthBearer.decode,
+    E.mapLeft(e =>
+      ResponseErrorValidation(
```
Comment (BurnedMarshal) [inline utils/auth_jwt.ts:88]: ditto

File: utils/auth_jwt.ts [hunk 5]
```diff
@@ -62,3 +69,36 @@ export const getValidateAuthJWT = ({
         TE.filterOrElse(AuthJWT.is, () => E.toError("Invalid AuthJWT payload"))
       )
   );
+
+/**
+ * A middleware that verify the jwt
```
Comment (BurnedMarshal) [inline utils/auth_jwt.ts:74]:
```suggestion
 * A middleware that verify the jwt and returns the payload
```

File: utils/auth_jwt.ts [hunk 6]
```diff
@@ -62,3 +69,36 @@ export const getValidateAuthJWT = ({
...
+ *
+ * @param jwtConfig the config for the jwt
+ * */
+
+export const verifyJWTMiddleware = (
+  jwtConfig: JWTConfig
+): IRequestMiddleware<"IResponseErrorValidation", AuthJWT> => (
```
Comment (BurnedMarshal) [inline utils/auth_jwt.ts:81]: `AuthJWT` should be configurable and not hardcoded in the method signature

File: utils/__tests__/auth_jwt.test.ts [hunk 7]
```diff
@@ -84,6 +94,120 @@ describe("getValidateJWT - Failures", () => {
...
+    WHEN VerifyJWTMiddleware is called\
+    THEN it should return a valid AuthJWT\
+    ", async () => {
+    const authJwt = await getGenerateAuthJWT(aConfigWithPrimaryKey)(aPayload)();
+    expect(E.isRight(authJwt)).toBeTruthy();
+
+    const middleware = verifyJWTMiddleware(aConfigWithTwoPrimaryKeys);
+
+    if (E.isRight(authJwt)) {
+      const req = {
```
Comment (BurnedMarshal) [inline utils/__tests__/auth_jwt.test.ts:109]:
```suggestion
      const mockReq = {
```

File: utils/__tests__/auth_jwt.test.ts [hunk 8]
```diff
@@ -84,6 +94,120 @@ describe("getValidateJWT - Failures", () => {
...
+        headers: {
+          "x-pagopa-lollipop-auth": `Bearer ${authJwt.right}`
+        }
+      } as any;
```
Comment (BurnedMarshal) [inline utils/__tests__/auth_jwt.test.ts:113]:
```suggestion
      } as unknown as Request;
```

File: utils/__tests__/auth_jwt.test.ts [hunk 9]
```diff
@@ -84,6 +94,120 @@ describe("getValidateJWT - Failures", () => {
...
+
+      expect(await middleware(req)).toMatchObject({
```
Comment (BurnedMarshal) [inline utils/__tests__/auth_jwt.test.ts:115]:
```
await expect(middleware(req)).resolve.toMatchObject({
```

File: utils/__tests__/auth_jwt.test.ts [hunk 10]
```diff
@@ -84,6 +94,120 @@ describe("getValidateJWT - Failures", () => {
...
+        _tag: "Right",
+        right: expect.objectContaining({
+          assertionRef: "sha256-p1NY7sl1d4lGvcTyYS535aZR_iJCleEIHFRE2lCHt-c",
+          operationId: "anOperationId",
+          iss: "test-issuer"
```
Comment (BurnedMarshal) [inline utils/__tests__/auth_jwt.test.ts:120]:
```suggestion
          iss: aConfigWithTwoPrimaryKeys.ISSUER
```

File: utils/__tests__/auth_jwt.test.ts [hunk 11]
```diff
@@ -84,6 +94,120 @@ describe("getValidateJWT - Failures", () => {
...
+    }
+  });
+
+  it("\
+    GIVEN a Valid jwtConfig and an empty x-pagopa-lollipop-auth\
+    WHEN VerifyJWTMiddleware is called\
+    THEN it should return a IResponseErrorValidation\
+    ", async () => {
+    const authJwt = await getGenerateAuthJWT(aConfigWithPrimaryKey)(aPayload)();
+    expect(E.isRight(authJwt)).toBeTruthy();
```
Comment (BurnedMarshal) [inline utils/__tests__/auth_jwt.test.ts:132]: not needed

File: utils/__tests__/auth_jwt.test.ts [hunk 12]
```diff
@@ -84,6 +94,120 @@ describe("getValidateJWT - Failures", () => {
...
+
+    const middleware = verifyJWTMiddleware(aConfigWithTwoPrimaryKeys);
+
+    if (E.isRight(authJwt)) {
```
Comment (BurnedMarshal) [inline utils/__tests__/auth_jwt.test.ts:136]: ditto

File: utils/__tests__/auth_jwt.test.ts [hunk 13]
```diff
@@ -84,6 +94,120 @@ describe("getValidateJWT - Failures", () => {
...
+  });
+
+  it("\
+    GIVEN a Valid jwtConfig and an invalid x-pagopa-lollipop-auth\
+    WHEN VerifyJWTMiddleware is called\
+    THEN it should return a IResponseErrorValidation containing LollipopAuthBearer decode failed\
+    ", async () => {
+    const authJwt = await getGenerateAuthJWT(aConfigWithPrimaryKey)(aPayload)();
+    const invalidAuth = "invalidAuth";
+    expect(E.isRight(authJwt)).toBeTruthy();
```
Comment (BurnedMarshal) [inline utils/__tests__/auth_jwt.test.ts:161]:
```suggestion
    const invalidAuth = "invalidAuth";
```

File: utils/__tests__/auth_jwt.test.ts [hunk 14]
```diff
@@ -84,6 +94,120 @@ describe("getValidateJWT - Failures", () => {
...
+  });
+
+  it("\
+    GIVEN a Valid jwtConfig and an x-pagopa-lollipop-auth valid ONLY for regex pattern\
+    WHEN VerifyJWTMiddleware is called\
+    THEN it should return a IResponseErrorValidation containing Invalid authJWT\
+    ", async () => {
+    const authJwt = await getGenerateAuthJWT(aConfigWithPrimaryKey)(aPayload)();
+    const invalidAuth = "Bearer aa";
+    expect(E.isRight(authJwt)).toBeTruthy();
```
Comment (BurnedMarshal) [inline utils/__tests__/auth_jwt.test.ts:189]:
```suggestion
    const invalidAuth = "Bearer aa";
```

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]

File: utils/auth_jwt.ts (hunk 6 above)
Comment (Garma00) [reply utils/auth_jwt.ts:81]: This would need a refactor of the `getValidateAuthJWT`

Comment (BurnedMarshal) [reply utils/auth_jwt.ts]: We'll make this change on another PR. Thanks.

**Review by BurnedMarshal — APPROVED** [review]

**Pull request merged**